File: algorithms/c++/RSelect.cpp
/* Uses a random partitioning algorithm RPartition to select the ith order statistic from
 * an array in expected linear time, with worst-case quadratic runtime.
 */

#include <random>
#include <iostream>
#include <cstdlib>
#include <vector>
using namespace std;

// We need a swap function
void Swap(vector<int> &A, int i, int j)
{
    int c = A[i];
    A[i] = A[j];
    A[j] = c;
}

// Partition using the last element as a pivot.
// Runtime: O(n)
int Partition(vector<int> &A, int p, int r)
{
    int i = p-1;
    for (int j=p; j<r; j++ )
    {
        if (A[j] < A[r])
        {
            i++;
            Swap(A, i, j);
        }
    }
    Swap(A, i+1, r);
    return i+1;
}

// Partition with a random pivot.
// Runtime: O(n)
int RPartition(vector<int> &A, int p, int r)
{
    random_device rd; // Should really seed once and then feed it to RPartition and RSelect
                      //    as an extra argument. That would probably save some time.
    mt19937 gen(rd());
    uniform_int_distribution<int> dis(p,r);
    Swap(A, r, dis(gen));
    return Partition(A, p, r);
}

// Now use RPartition to get the ith order statistic of the subarray A[p..r].
// Expected runtime: <O(n)>; Worst case: <O(n^2)>
int RSelect(vector<int> &A, int p, int r, int i)
{
    if (p==r) return A[p];
    int q = RPartition(A, p, r);
    int k = q-p+1;
    if (k==i) return A[q];
    else if (i<k) return RSelect(A, p, q-1, i);
    else return RSelect(A, q+1, r, i-k);
}

// now let's see if it works
int main(int argc, char **argv)
{
    vector<int> A(argc-2); // make room for all args except first and last
    for (int i=0; i<argc-2; i++) A[i] = atoi(argv[i+1]); // get all args except first and last
    int i = atoi(argv[argc-1]); // the last arg is the order 
    cout << RSelect(A, 0, A.size()-1, i) << endl;
    return 0;
}

File: algorithms/python/RSelect.py
# Uses a random partitioning algorithm RPartition to select the ith order statistic (i.e.
#   the ith smallest value) of an array
#   in expected linear time, with worst-case quadratic runtime.

# Partition the subarray A[p..r] of a around a random element. Elements strictly less than the
#   pivot go to the left.
# Runtime: O(n)

# Regular partition: partitions around the last element.
# Runtim: O(n)
def Partition(A, p, r):
    i = p-1
    for j in range(p,r):
        if A[j] < A[r]:
            i+=1
            A[j],A[i] = A[i],A[j]
    A[r],A[i+1] = A[i+1],A[r]
    return i+1

# Random partition:
# Runtime: O(n)
from random import random
def RPartition(A, p, r):
    i = p + int(random()*(r-p+1)) # pick a random index from [p..r]
    A[i],A[r] = A[r],A[i]
    return Partition(A, p, r)

# Recursively select from the side of the partition where we know the ith order statistic
#   must be. Recursion stops when the pivot itself is the ith order statistic, or when Select
#   sees a length 1 list.
# Expected Runtime: <O(n)>; Worst case: <O(n^2)>
def RSelect(A, p, r, i):
    if p==r:
        return A[p]     # we've recursed down to a length 1 list
    q = RPartition(A, p, r)
    k = q-p+1
    if k==i:
        return A[q]     # the pivot value is the answer
    elif i<k:
        return RSelect(A, p, q-1, i)
    else:
        return RSelect(A, q+1, r, i-k)

